### PR TITLE
Made the session optional for database with  replica set not configured.

### LINKED ISSUE #1772: Edit Setup.ts to reflect database needs
**Describe the bug**
Currently, Talawa-api needs to have access to either a local database with a replica set or mongoDB atlas to use the retry writes option. Our setup.ts script does not support the user in ensure that one of these options is met. 

**Expected behavior**
Setup.ts needs to guide the user to make sure the database will work with retry rewrites. 

**Actual behavior**
Setup does not give the user this guidance. 

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by adi790uu**
**What kind of change does this PR introduce?**

bugfix

**Issue Number:**

Fixes #1772 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

https://github.com/PalisadoesFoundation/talawa-api/assets/132701661/726039a8-72d9-4afc-9e4f-72a9aced1700

Screenshot 2024-02-11 at 10 39 15 PM

**If relevant, did you update the documentation?**

No

**Summary**

Made the session optional for local database which does not have replica set enabled.

**Does this PR introduce a breaking change?**

No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
This solution only works with `mongodb.net`

1. How can you make this universally applicable?
2. How can we also fix this with `setup.ts`?

**Comment by palisadoes** [comment]
@EshaanAgg, @varshith257, @meetulr

1. You all had opinions on the approach. Please take a look here. My concern is that this solution isn't universally applicable as it checks or the `mongodb.net` domain.
2. It would seem that we'd need to do some connectivity test to be certain, and possibly have modifications to `setup.ts`

**Comment by meetulr** [comment]
I was just thinking of running the [`hello`](https://www.mongodb.com/docs/v4.4/reference/command/hello/#:~:text=MongoDB%20drivers%20and%20clients%20use%20hello%20to%20determine,following%20form%3A%20db.%20runCommand%20%28%7B%20hello%3A%201%20%7D%29) command.

It returns information about the instance. We can check if it has properties that only a replica set would have (like hosts or setName), and start a session if it does.

**Comment by adi790uu** [comment]
> I was just thinking of running the [`hello`](https://www.mongodb.com/docs/v4.4/reference/command/hello/#:~:text=MongoDB%20drivers%20and%20clients%20use%20hello%20to%20determine,following%20form%3A%20db.%20runCommand%20%28%7B%20hello%3A%201%20%7D%29) command.
> 
> It returns information about the instance. We can check if it has properties that only a replica set would have (like hosts or setName), and start a session if it does.

Yeah this seems to be a valid approach.

**Comment by varshith257** [comment]
@palisadoes I agree with @meetulr to make the solution universally applicable, we could perform a connectivity test to check if the MongoDB instance is a replica set, instead of checking for the `mongodb.net` domain. This can be done by running the `hello` command and checking if the instance has properties that only a replica set would have, such as hosts or setName. If it does, then we can start a session.

**Comment by varshith257** [comment]
Regarding `setup.ts`, we can add a warning message to inform the user if they have provided a local MongoDB URL with retry writes enabled. However, the ultimate solution would be to make the session optional for local databases without a replica set.

**Comment by varshith257** [comment]
> @EshaanAgg, @varshith257, @meetulr
> 
> 1. You all had opinions on the approach. Please take a look here. My concern is that this solution isn't universally applicable as it checks or the `mongodb.net` domain.
> 2. It would seem that we'd need to do some connectivity test to be certain, and possibly have modifications to `setup.ts`

Regarding `setup.ts`, we can add a warning message to inform the user if they have provided a local MongoDB URL with retry writes enabled. However, the ultimate solution would be to make the session optional for local databases without a replica set.

We can implement a check for it in the `db.ts` . If the MongoDB instance is not a replica set, the session will not be initiated. This approach will ensure that we don't face any issues with local single-node standalone instances, and we can still have support for transactions in deployed versions.

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by palisadoes** [comment]
@adi790uu How's progress on this?

**Comment by adi790uu** [comment]
@palisadoes almost done will create a PR today.

**Comment by adi790uu** [comment]
hey @palisadoes I am getting es-lint errors after I merged the latest code from develop branch to the pr branch

**Comment by adi790uu** [comment]
@palisadoes @Cioppolo14 any suggestions to solve above there are many linting errors so it is impossible to resolve them.

**Comment by varshith257** [comment]
> @palisadoes @Cioppolo14 any suggestions to solve above there are many linting errors so it is impossible to resolve them.
> 

You can make use of `lint:fix` command probably

**Comment by adi790uu** [comment]
> > @palisadoes @Cioppolo14 any suggestions to solve above there are many linting errors so it is impossible to resolve them.
> 
> You can make use of `lint:fix` command probably

already tried that

**Comment by palisadoes** [comment]
> hey @palisadoes I am getting es-lint errors after I merged the latest code from develop branch to the pr branch. Maybe I have to create a new branch out of develop ,make my changes and then create a new pr for this issue to avoid this error.

Try that, but don't close this PR. Try deleting your local branch and recreating a branch with the same name from develop.  Make your changes before committing. You may have to do a force push to your origin too.

**Comment by adi790uu** [comment]
@palisadoes I have not added any warning in setup.ts regarding replica set configuration should I do it but I think its not required for local development purpose.

**Comment by palisadoes** [comment]
1. That should be OK.
2. @meetulr, @varshith257 Is this sufficient?

**Comment by meetulr** [comment]
> 1. That should be OK.
> 2. @meetulr, @varshith257 Is this sufficient?

This is what I did... i.e. using the `hello` command.

Would be better if we added some logs & comments in here though, and check for a property that's more intuitive to understand, like `setName` (if it's a replica set, it would have a `setName`, otherwise it won't). Something like this:

connect-function

connect-function
non-replica-log
replica-set-log

**Comment by adi790uu** [comment]
> > 1. That should be OK.
> > 2. @meetulr, @varshith257 Is this sufficient?
> 
> This is what I did... i.e. using the `hello` command.
> 
> Would be better if we added some logs & comments in here. And check for a property that's more intuitive to understand, like `setName` (if it's a replica set, it would have a `setName`, otherwise it won't). Something like this:
> 
> connect-function
> checkIfReplicaSet
> 
> non-replica-log
> replica-set-log

Yeah we can do logging and in my code I am checking for hosts array if it's present then it's a replica set otherwise it's not.

**Comment by adi790uu** [comment]
@palisadoes is logging required, should I implement that?

**Comment by varshith257** [comment]
> @palisadoes is logging required, should I implement that?

Go for it. It indicates info of session

**Comment by palisadoes** [comment]
If the type of logging does not significantly degrade performance, and is available for the community versions of Mongo, then we should do it.

**Comment by adi790uu** [comment]
@palisadoes implemented logging and I have written test as well for the script, do give your feedback if any change is still required.

**Comment by palisadoes** [comment]
Have you tested this with both a local and cloud instances? We must be able to create events without issue in Talawa-Admin

**Comment by adi790uu** [comment]
I have tested it locally @palisadoes

**Comment by palisadoes** [comment]
1. Using the mongoDB Atlas cloud service which many of the developers seem to be using. It uses replica sets.
2. Ask one of the others in this PR's thread and that of the original issue to test against this PR too.

**Comment by adi790uu** [comment]
> 1. Using the mongoDB Atlas cloud service which many of the developers seem to be using. It uses replica sets.
> 2. Ask one of the others in this PR's thread and that of the original issue to test against this PR too.

Yeah I have tested the mongdb atlas instance, I misunderstood the question.

**Comment by adi790uu** [comment]
@varshith257 @meetulr can you verify/review this

**Review by varshith257 — CHANGES_REQUESTED** [review]
As I suggested earlier, Can you add an additional check for the 'setName' property in the replica set condition?

File: src/db.ts [hunk 1]
```diff
@@ -12,7 +12,14 @@ export const connect = async (): Promise<void> => {
       useFindAndModify: false,
       useNewUrlParser: true,
     });
-    session = await mongoose.startSession();
+
+    const adminDb = mongoose.connection.db.admin();
+    const result = await adminDb.command({
+      hello: 1,
+    });
+    if ("hosts" in result) {
```
[the remark this answers is not in the document]
Comment (adi790uu) [reply src/db.ts]: Yeah we can do that, ok will add that as well.

**Review by EshaanAgg — DISMISSED** [review]

File: src/db.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (palisadoes) [reply src/db.ts]: Yes, add this additional check so we can merge the PR.

**Review by varshith257 — APPROVED** [review]
WhatsApp Image 2024-02-20 at 00 36 57_9e9feb7d

LGTM

**Comment by adi790uu** [comment]
@palisadoes added the setName check as well

**Review by EshaanAgg — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Pull request merged**